### PR TITLE
Add rule S1862 Related "if/else if" statements should not have the same condition

### LINKED ISSUE #40: Rule S1862: Related "if/else if" statements should not have the same condition
[RSPEC-1862](https://jira.sonarsource.com/browse/RSPEC-1862)

### PR DESCRIPTION
**Pull request by saberduck**
Fix #40

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/IfLike.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return node;
+  }
+
+  public UastNode condition() {
+    return condition;
+  }
+
+  @CheckForNull
+  public UastNode elseNode() {
+    return node.getChild(UastNode.Kind.ELSE).orElse(null);
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/helpers/IfLike.java]: Would it make sense to create an `elseNode` field initialized at construction-time like `condition`?

File: common-rule-engine/src/test/resources/org/sonar/commonruleengine/checks/go/NoIdenticalConditionsCheck/NoIdenticalConditionsCheck.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	if condition1 {
+	} else if condition2 {
+	} else if condition1 { // Noncompliant [[secondary=12]] {{This branch can not be reached because the condition duplicates a previous condition in the same sequence of "if/else if" statements}}
+	}
+}
+
+func ifInitializer(x int)  {
+	if x = 3;x > 0 {
+
+	} else if x = -1;x > 0 { // Compliant, initializer has to be considered as part of condition
```
Comment (ghost) [inline common-rule-engine/src/test/resources/org/sonar/commonruleengine/checks/go/NoIdenticalConditionsCheck/NoIdenticalConditionsCheck.go]: Missing space after `;`. I know this is just a dummy example, but I still find it easier to focus on the meat of it without distractions.

**Review by ghost — APPROVED** [review]

**Comment by saberduck** [comment]
@janos-ss can you have another look on handling of `switch` , note that `switch` part is not implemented in Java because this is not allowed by language.

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalConditionsCheck.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  public NoIdenticalConditionsCheck() {
+    super(UastNode.Kind.IF, UastNode.Kind.SWITCH);
+  }
+
+  @Override
+  public void visitNode(UastNode node) {
+    if (node.kinds.contains(UastNode.Kind.IF)) {
+      handleIf(node);
+    }
+    if (node.kinds.contains(UastNode.Kind.SWITCH)) {
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/NoIdenticalConditionsCheck.java]: Could it ever make sense for a node to have both IF and SWITCH kind at the same time?

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java [hunk 4]
```diff
@@ -32,12 +33,8 @@ public UastNode body() {
...
 
   public List<UastNode> parameters() {
-    List<UastNode> parameters = node.getChildren(UastNode.Kind.PARAMETER);
-    if (parameters.isEmpty()) {
-      node.children.stream()
-        .filter(child -> !child.kinds.contains(UastNode.Kind.BLOCK))
-        .forEach(child -> child.getDescendants(UastNode.Kind.PARAMETER, parameters::add));
-    }
+    List<UastNode> parameters = new ArrayList<>();
+    node.getDescendants(UastNode.Kind.PARAMETER, parameters::add);
```
Comment (ghost) [inline common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java]: If we don't exclude `BLOCK`, and the function body contains some function literals with params, those params will be incorrectly included.

File: common-rule-engine/src/test/resources/org/sonar/commonruleengine/checks/go/NoIdenticalConditionsCheck/NoIdenticalConditionsCheckSwitch.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	case param > 1: // Noncompliant
+		fmt.Println("impossible")
+	}
+}
+
+func SwitchWithMultipleConditions(param int) {
+	switch param {
+	case 1, 2, 3:
+		fmt.Println(">1")
+	case 3, 4, 5: // Noncompliant
```
Comment (ghost) [inline common-rule-engine/src/test/resources/org/sonar/commonruleengine/checks/go/NoIdenticalConditionsCheck/NoIdenticalConditionsCheckSwitch.go]: It was not obvious to me at first why this is noncompliant. (Duplicate 3)

File: uast-generator-go/goparser.go [hunk 6]
```diff
@@ -242,14 +254,56 @@ func mapStmt(astNode ast.Stmt) *Node {
 	return node
 }
 
+func mapCaseClause(clause *ast.CaseClause) *Node {
+	var children []*Node
+	children = append(children, mapToken(token.CASE, clause.Case))
```
Comment (ghost) [inline uast-generator-go/goparser.go]: Two lines could be one with: `children := children(mapToken(token.CASE, clause.Case))`

File: uast-generator-go/goparser.go [hunk 7]
```diff
@@ -242,14 +254,56 @@ func mapStmt(astNode ast.Stmt) *Node {
...
+	children = append(children, childrenFromNodeList(StmtList(clause.Body))...)
+	return &Node{
+		Kinds:    kinds(clause),
+		Children: children,
+	}
+}
+
+func mapSwitchStmt(stmt *ast.SwitchStmt) *Node {
+	var children []*Node
+	children = append(children, mapToken(token.SWITCH, stmt.Switch))
```
Comment (ghost) [inline uast-generator-go/goparser.go]: Two lines could be written in one, as in my earlier comment

File: uast-generator-go/goparser_test.go [hunk 8]
```diff
@@ -317,6 +317,42 @@ func main() {
...
+}
+`
+	_, astFile := astFromString(source)
+	blockStmt := astFile.Decls[0].(*ast.FuncDecl).Body
+	switchStmt := blockStmt.List[0].(*ast.SwitchStmt)
+	conditionCounter := 0
+	for _, caseNode := range switchStmt.Body.List {
+		caseClause := mapCaseClause(caseNode.(*ast.CaseClause))
+		expectKinds(t, caseClause, kinds(CASE))
+		for _, children := range caseClause.Children {
```
Comment (ghost) [inline uast-generator-go/goparser_test.go]: Singular `child` would be more natural ;-)

File: uast-generator-go/goparser_test.go [hunk 9]
```diff
@@ -317,6 +317,42 @@ func main() {
...
+			if children.hasKind(CONDITION) {
+				conditionCounter++
+			}
+		}
+	}
+	if conditionCounter != 6 {
+		t.Fatalf("got %v conditions; expected %v", 6, conditionCounter)
```
Comment (ghost) [inline uast-generator-go/goparser_test.go]: The parameters are in reverse order (6 should be the last, or just hardcoded in the string)

**Review by ghost — APPROVED** [review]
LGTM with the change discussed about descendants, and some minor optional changes (keep in mind we're going to replace this Go parser)

File: common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java (hunk 4 above)
Comment (saberduck) [reply common-rule-engine/src/main/java/org/sonar/uast/helpers/FunctionLike.java]: Indeed

**Pull request merged**